File: Laptop.cs
using System;
using System.Collections.Generic;

class Program
{
    static void Main(string[] args)
    {
        LaptopShop Ls = new LaptopShop();
        Ls.SetLaptopShop("Ronaldo.Shop");
        Ls.LoadData();
        Console.WriteLine(Ls);

        LaptopShop Ls1 = Ls.Clone();
        Ls.GetLaptops().RemoveAt(2);
        Ls1.SetLaptopShop("Messi");
        Console.WriteLine(Ls1);
    }
}

class LaptopShop : ICloneable
{
    private string laptopShop;
    private List<Laptop> laptops = new List<Laptop>();
    private Random random = new Random();

    public string GetLaptopShop()
    {
        return laptopShop;
    }
    public void SetLaptopShop(string shopName)
    {
        laptopShop = shopName;
    }
    public List<Laptop> GetLaptops()
    {
        return laptops;
    }
    public void LoadData()
    {
        for (int i = 0; i < 10; i++)
        {
            Laptop l = new Laptop();
            l.SetLap_ID(i);
            string temp = (random.Next(0, 10) > 5) ? "Ryzen" : "Intel";
            l.SetLap_processor(temp);
            laptops.Add(l);
        }
    }
    public override string ToString()
    {
        return "Shop=[Shopname=" + laptopShop + ",Laptops=" + string.Join(",", laptops) + "]";
    }
    public object Clone()
    {
        LaptopShop shop = new LaptopShop();
        shop.SetLaptopShop(laptopShop);

        foreach (Laptop l in laptops)
        {
            shop.GetLaptops().Add(l.Clone() as Laptop);
        }

        return shop;
    }
}

class Laptop : ICloneable
{
    private int Lap_ID;
    private string Lap_processor;

    public int GetLap_ID()
    {
        return Lap_ID;
    }
    public void SetLap_ID(int id)
    {
        Lap_ID = id;
    }
    public string GetLap_processor()
    {
        return Lap_processor;
    }
    public void SetLap_processor(string processor)
    {
        Lap_processor = processor;
    }

    public override string ToString()
    {
        return "{Lap_ID=" + Lap_ID + ", Lap_processor=" + Lap_processor + "}";
    }
    public object Clone()
    {
        return this.MemberwiseClone();
    }
}

File: Laptop.cpp
#include <iostream>
#include <vector>
#include <string>
#include <random>
#include <memory>

class Laptop;

class LaptopShop {
private:
    std::string laptopShop;
    std::vector<std::shared_ptr<Laptop>> laptops;
    std::random_device rd;
    std::mt19937 gen{rd()};
    std::uniform_int_distribution<> dis{0, 10};

public:
    LaptopShop() = default;
    LaptopShop(const std::string& shopName) : laptopShop(shopName) {}

    std::string getLaptopShop() const {
        return laptopShop;
    }

    void setLaptopShop(const std::string& shopName) {
        laptopShop = shopName;
    }

    const std::vector<std::shared_ptr<Laptop>>& getLaptops() const {
        return laptops;
    }

    void loadData() {
        for (int i = 0; i < 10; i++) {
            auto l = std::make_shared<Laptop>();
            l->setLap_ID(i);
            std::string temp = (dis(gen) > 5) ? "Ryzen" : "Intel";
            l->setLap_processor(temp);
            laptops.push_back(l);
        }
    }

    friend std::ostream& operator<<(std::ostream& os, const LaptopShop& shop) {
        os << "Shop=[Shopname=" << shop.laptopShop << ",Laptops=[";
        for (const auto& lap : shop.laptops) {
            os << "{Lap_ID=" << lap->getLap_ID() << ", Lap_processor=" << lap->getLap_processor() << "}, ";
        }
        os << "]]";
        return os;
    }

    LaptopShop clone() const {
        LaptopShop shop;
        shop.setLaptopShop(laptopShop);
        for (const auto& l : laptops) {
            auto l_clone = std::make_shared<Laptop>();
            l_clone->setLap_ID(l->getLap_ID());
            l_clone->setLap_processor(l->getLap_processor());
            shop.laptops.push_back(l_clone);
        }
        return shop;
    }
};

class Laptop {
private:
    int Lap_ID;
    std::string Lap_processor;

public:
    int getLap_ID() const {
        return Lap_ID;
    }

    void setLap_ID(int id) {
        Lap_ID = id;
    }

    const std::string& getLap_processor() const {
        return Lap_processor;
    }

    void setLap_processor(const std::string& processor) {
        Lap_processor = processor;
    }
};

int main() {
    LaptopShop Ls("Ronaldo.Shop");
    Ls.loadData();
    std::cout << Ls << std::endl;

    LaptopShop Ls1 = Ls.clone();
    Ls.getLaptops().pop_back();
    Ls1.setLaptopShop("Messi");
    std::cout << Ls1 << std::endl;

    return 0;
}
